File: ios/AgenticInbox/Inboxies/Views/Settings/SendersSettingsView.swift
import SwiftUI

/// Per-sender purpose-box defaults (Inbox / Promotions / Updates).
struct SendersSettingsView: View {
    @Environment(AppModel.self) private var app
    @Environment(\.dismiss) private var dismiss

    @State private var preferences: [SenderPreference] = []
    @State private var query = ""
    @State private var isLoading = true
    @State private var errorMessage: String?
    @State private var busyAddress: String?

    private static let purposeFolders: [(id: String, name: String)] = [
        ("inbox", "Inbox"),
        ("promotions", "Promotions"),
        ("updates", "Updates"),
    ]

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                Text("Choose which purpose box each sender goes to. Changing a destination also moves their existing Inbox, Promotions, and Updates mail.")
                    .font(.inter(size: 13))
                    .foregroundStyle(AppTheme.muted)
                    .padding(.horizontal, 20)

                TextField("Search senders", text: $query)
                    .font(.inter(size: 15))
                    .padding(.horizontal, 14)
                    .padding(.vertical, 10)
                    .background(AppTheme.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
                    .padding(.horizontal, 20)
                    .onChange(of: query) { _, _ in
                        Task { await load() }
                    }

                if isLoading {
                    ProgressView()
                        .frame(maxWidth: .infinity)
                        .padding(.top, 24)
                } else if let errorMessage {
                    Text(errorMessage)
                        .font(.inter(size: 13))
                        .foregroundStyle(Color.red.opacity(0.85))
                        .padding(.horizontal, 20)
                } else if preferences.isEmpty {
                    VStack(alignment: .leading, spacing: 6) {
                        Text("No sender defaults yet")
                            .font(.inter(size: 15, weight: .medium))
                            .foregroundStyle(AppTheme.ink)
                        Text("When you move mail into Inbox, Promotions, or Updates, you can set where future mail from that sender goes.")
                            .font(.inter(size: 13))
                            .foregroundStyle(AppTheme.muted)
                    }
                    .padding(.horizontal, 20)
                    .padding(.top, 8)
                } else {
                    VStack(spacing: 0) {
                        ForEach(preferences) { pref in
                            senderRow(pref)
                            if pref.id != preferences.last?.id {
                                Divider().overlay(AppTheme.line)
                            }
                        }
                    }
                    .background(AppTheme.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                    .padding(.horizontal, 20)
                }

                Text("Filters still override these defaults when a rule matches. Spam is never routed by sender preference.")
                    .font(.inter(size: 12))
                    .foregroundStyle(AppTheme.muted)
                    .padding(.horizontal, 20)
                    .padding(.bottom, 24)
            }
            .padding(.top, 8)
        }
        .background(AppTheme.background)
        .navigationTitle("Senders")
        .navigationBarTitleDisplayMode(.inline)
        .task { await load() }
    }

    @ViewBuilder
    private func senderRow(_ pref: SenderPreference) -> some View {
        HStack(spacing: 12) {
            VStack(alignment: .leading, spacing: 2) {
                Text(pref.displayName?.isEmpty == false ? pref.displayName! : pref.address)
                    .font(.inter(size: 15, weight: .medium))
                    .foregroundStyle(AppTheme.ink)
                    .lineLimit(1)
                if let name = pref.displayName, !name.isEmpty {
                    Text(pref.address)
                        .font(.inter(size: 12))
                        .foregroundStyle(AppTheme.muted)
                        .lineLimit(1)
                }
            }
            Spacer(minLength: 8)
            Menu {
                ForEach(Self.purposeFolders, id: \.id) { folder in
                    Button(folder.name) {
                        Task { await updateFolder(pref, folderId: folder.id) }
                    }
                }
            } label: {
                Text(Self.purposeFolders.first { $0.id == pref.folderId }?.name ?? pref.folderId)
                    .font(.inter(size: 13, weight: .medium))
                    .foregroundStyle(AppTheme.accent)
            }
            .disabled(busyAddress == pref.address)

            Button {
                Task { await remove(pref) }
            } label: {
                Image(systemName: "trash")
                    .font(.system(size: 14))
                    .foregroundStyle(AppTheme.muted)
            }
            .disabled(busyAddress == pref.address)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }

    private func load() async {
        guard let mailboxId = app.selectedMailboxId else {
            isLoading = false
            return
        }
        isLoading = preferences.isEmpty
        errorMessage = nil
        do {
            preferences = try await APIClient.shared.listSenderPreferences(
                mailboxId: mailboxId,
                q: query.trimmingCharacters(in: .whitespacesAndNewlines)
            )
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    private func updateFolder(_ pref: SenderPreference, folderId: String) async {
        guard let mailboxId = app.selectedMailboxId else { return }
        busyAddress = pref.address
        defer { busyAddress = nil }
        do {
            _ = try await APIClient.shared.upsertSenderPreference(
                mailboxId: mailboxId,
                address: pref.address,
                folderId: folderId,
                displayName: pref.displayName,
                refile: true
            )
            await load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func remove(_ pref: SenderPreference) async {
        guard let mailboxId = app.selectedMailboxId else { return }
        busyAddress = pref.address
        defer { busyAddress = nil }
        do {
            try await APIClient.shared.deleteSenderPreference(
                mailboxId: mailboxId,
                address: pref.address
            )
            preferences.removeAll { $0.address == pref.address }
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: android/app/src/main/java/co/inboxies/app/ui/settings/SendersSettingsView.kt
package co.inboxies.app.ui.settings

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import co.inboxies.app.LocalAppModel
import co.inboxies.app.models.FolderIds
import co.inboxies.app.models.SenderPreference
import co.inboxies.app.services.ApiClient
import co.inboxies.app.theme.HomeChromeMetrics
import co.inboxies.app.theme.HomeChromeToolbarButton
import co.inboxies.app.theme.InterFontFamily
import co.inboxies.app.theme.inboxiesColors
import kotlinx.coroutines.launch

@Composable
fun SendersSettingsView(
    onBack: () -> Unit,
) {
    val app = LocalAppModel.current
    val colors = inboxiesColors()
    val scope = rememberCoroutineScope()
    val mailboxId = app.selectedMailbox?.id

    var query by remember { mutableStateOf("") }
    var preferences by remember { mutableStateOf<List<SenderPreference>>(emptyList()) }
    var isLoading by remember { mutableStateOf(true) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var busyAddress by remember { mutableStateOf<String?>(null) }

    fun reload() {
        val id = mailboxId ?: return
        scope.launch {
            isLoading = preferences.isEmpty()
            errorMessage = null
            runCatching {
                ApiClient.shared.listSenderPreferences(id, q = query.trim())
            }.onSuccess {
                preferences = it
            }.onFailure {
                errorMessage = it.message ?: "Failed to load senders"
            }
            isLoading = false
        }
    }

    LaunchedEffect(mailboxId, query) {
        reload()
    }

    Box(modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.fillMaxWidth()) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 8.dp, vertical = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                HomeChromeToolbarButton(
                    icon = Icons.AutoMirrored.Outlined.ArrowBack,
                    contentDescription = "Back",
                    onClick = onBack,
                )
                Text(
                    "Senders",
                    fontFamily = InterFontFamily,
                    fontSize = 17.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = colors.ink,
                    modifier = Modifier.padding(start = 4.dp),
                )
            }

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = 20.dp, vertical = 12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Text(
                    "Choose which purpose box each sender goes to. Changing a destination also moves their existing Inbox, Promotions, and Updates mail.",
                    fontFamily = InterFontFamily,
                    fontSize = 13.sp,
                    color = colors.muted,
                )

                OutlinedTextField(
                    value = query,
                    onValueChange = { query = it },
                    modifier = Modifier.fillMaxWidth(),
                    singleLine = true,
                    placeholder = { Text("Search senders") },
                )

                when {
                    isLoading -> {
                        Box(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(vertical = 24.dp),
                            contentAlignment = Alignment.Center,
                        ) {
                            CircularProgressIndicator(
                                color = colors.accent,
                                modifier = Modifier.size(24.dp),
                                strokeWidth = 2.dp,
                            )
                        }
                    }
                    errorMessage != null -> {
                        Text(
                            errorMessage!!,
                            fontFamily = InterFontFamily,
                            fontSize = 13.sp,
                            color = colors.deepDarkRed,
                        )
                    }
                    preferences.isEmpty() -> {
                        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                            Text(
                                "No sender defaults yet",
                                fontFamily = InterFontFamily,
                                fontSize = 15.sp,
                                fontWeight = FontWeight.Medium,
                                color = colors.ink,
                            )
                            Text(
                                "When you move mail into Inbox, Promotions, or Updates, you can set where future mail from that sender goes.",
                                fontFamily = InterFontFamily,
                                fontSize = 13.sp,
                                color = colors.muted,
                            )
                        }
                    }
                    else -> {
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(colors.surface, RoundedCornerShape(12.dp)),
                        ) {
                            preferences.forEachIndexed { index, pref ->
                                SenderPreferenceRow(
                                    preference = pref,
                                    busy = busyAddress == pref.address,
                                    onChangeFolder = { folderId ->
                                        val id = mailboxId ?: return@SenderPreferenceRow
                                        busyAddress = pref.address
                                        scope.launch {
                                            runCatching {
                                                ApiClient.shared.upsertSenderPreference(
                                                    id,
                                                    pref.address,
                                                    folderId,
                                                    displayName = pref.displayName,
                                                    refile = true,
                                                )
                                            }.onSuccess {
                                                reload()
                                            }.onFailure {
                                                errorMessage = it.message
                                            }
                                            busyAddress = null
                                        }
                                    },
                                    onDelete = {
                                        val id = mailboxId ?: return@SenderPreferenceRow
                                        busyAddress = pref.address
                                        scope.launch {
                                            runCatching {
                                                ApiClient.shared.deleteSenderPreference(id, pref.address)
                                            }.onSuccess {
                                                preferences = preferences.filterNot { it.address == pref.address }
                                            }.onFailure {
                                                errorMessage = it.message
                                            }
                                            busyAddress = null
                                        }
                                    },
                                )
                                if (index != preferences.lastIndex) {
                                    Spacer(
                                        modifier = Modifier
                                            .fillMaxWidth()
                                            .height(1.dp)
                                            .background(colors.line),
                                    )
                                }
                            }
                        }
                    }
                }

                Text(
                    "Filters still override these defaults when a rule matches. Spam is never routed by sender preference.",
                    fontFamily = InterFontFamily,
                    fontSize = 12.sp,
                    color = colors.muted,
                )
                Spacer(modifier = Modifier.height(24.dp))
            }
        }
    }
}

@Composable
private fun SenderPreferenceRow(
    preference: SenderPreference,
    busy: Boolean,
    onChangeFolder: (String) -> Unit,
    onDelete: () -> Unit,
) {
    val colors = inboxiesColors()
    var menuOpen by remember { mutableStateOf(false) }
    val title = preference.displayName?.takeIf { it.isNotBlank() } ?: preference.address

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 14.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                title,
                fontFamily = InterFontFamily,
                fontSize = 15.sp,
                fontWeight = FontWeight.Medium,
                color = colors.ink,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            if (!preference.displayName.isNullOrBlank()) {
                Text(
                    preference.address,
                    fontFamily = InterFontFamily,
                    fontSize = 12.sp,
                    color = colors.muted,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }

        Box {
            Text(
                FolderIds.purposeDisplayName(preference.folderId),
                fontFamily = InterFontFamily,
                fontSize = 13.sp,
                fontWeight = FontWeight.Medium,
                color = colors.accent,
                modifier = Modifier
                    .clickable(enabled = !busy) { menuOpen = true }
                    .padding(vertical = 4.dp, horizontal = 4.dp),
            )
            DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                FolderIds.purposeFolderIds.forEach { folderId ->
                    DropdownMenuItem(
                        text = { Text(FolderIds.purposeDisplayName(folderId)) },
                        onClick = {
                            menuOpen = false
                            onChangeFolder(folderId)
                        },
                    )
                }
            }
        }

        Icon(
            Icons.Outlined.Delete,
            contentDescription = "Remove",
            tint = colors.muted,
            modifier = Modifier
                .size(18.dp)
                .clickable(enabled = !busy, onClick = onDelete),
        )
    }
}
